crates/services/upgradable-executor/wasm-executor/src/utils.rs

```diff
@@ -80,16 +84,166 @@ pub enum WasmDeserializationBlockTypes<TxSource> {
     Validation(Block),
 }
 
+/// The JSON version of the executor error. The serialization and deserialization
+/// of the JSON error are less sensitive to the order of the variants in the enum.
+/// It simplifies the error conversion between different versions of the execution.
+///
+/// If deserialization fails, it returns a string representation of the error that
+/// still has useful information, even if the error is not supported by the native executor.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct JSONError(String);
+
+#[cfg(feature = "std")]
+impl From<ExecutorErrorV0> for JSONError {
+    fn from(value: ExecutorErrorV0) -> Self {
+        let json = serde_json::to_string(&value).unwrap_or_else(|e| {
+            anyhow::anyhow!("Failed to serialize the V0 error: {:?}", e).to_string()
+        });
+        JSONError(json)
+    }
+}
+
+impl From<ExecutorError> for JSONError {
+    fn from(value: ExecutorError) -> Self {
+        let json = serde_json::to_string(&value).unwrap_or_else(|e| {
+            anyhow::anyhow!("Failed to serialize the error: {:?}", e).to_string()
+        });
+        JSONError(json)
+    }
+}
+
+impl From<JSONError> for ExecutorError {
+    fn from(value: JSONError) -> Self {
+        serde_json::from_str(&value.0).unwrap_or(ExecutorError::Other(value.0))
+    }
+}
+
 /// The return type for the WASM executor. Enum allows handling different
 /// versions of the return without introducing new host functions.
+#[cfg(feature = "std")]
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
 pub enum ReturnType {
-    V1(ExecutorResult<Uncommitted<ExecutionResult, Changes>>),
+    ExecutionV0(
+        Result<Uncommitted<ExecutionResult<ExecutorErrorV0>, Changes>, ExecutorErrorV0>,
+    ),
+    ExecutionV1(Result<Uncommitted<ExecutionResult<JSONError>, Changes>, JSONError>),
+    Validation(Result<Uncommitted<ValidationResult, Changes>, JSONError>),
+}
+
+/// The return type for the WASM executor. Enum allows handling different
+/// versions of the return without introducing new host functions.
+#[cfg(not(feature = "std"))]
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub enum ReturnType {
+    ExecutionV0,
+    ExecutionV1(Result<Uncommitted<ExecutionResult<JSONError>, Changes>, JSONError>),
+    Validation(Result<Uncommitted<ValidationResult, Changes>, JSONError>),
+}
+
+/// Converts the latest execution result to the v1 execution result.
```

Comment:
what's the difference between the latest execution result and the v1 execution result? (I thought v1 is the latest version)